ql/src/java/org/apache/hadoop/hive/ql/queryhistory/persist/AbstractQueryHistoryPersistor.java

```diff
@@ -0,0 +1,138 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.queryhistory.persist;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.TableType;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.metastore.api.hive_metastoreConstants;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.parse.PartitionTransform;
+import org.apache.hadoop.hive.ql.parse.TransformSpec;
+import org.apache.hadoop.hive.ql.queryhistory.schema.QueryHistorySchema;
+import org.apache.hadoop.hive.ql.session.SessionStateUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public abstract class AbstractQueryHistoryPersistor implements QueryHistoryPersistor {
+  protected Logger LOG = LoggerFactory.getLogger(getClass());
+  @VisibleForTesting
+  HiveConf conf;
+  protected QueryHistorySchema schema;
+
+  public void init(HiveConf conf, QueryHistorySchema schema) {
+    this.conf = conf;
+    this.schema = schema;
+
+    Database database = initDatabase();
+    Table table = initTable(database);
+    try {
+      postInitTable(table);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected Database initDatabase() {
+    try (Hive hive = Hive.get(conf)) {
+      Database db = null;
+      try {
+        db = hive.getDatabase(QUERY_HISTORY_DB_NAME);
+        if (db == null) {
+          LOG.warn("Database ({}) for query history table hasn't been found, auto-creating one", QUERY_HISTORY_DB_NAME);
+          String location = getDatabaseLocation(QueryHistoryPersistor.QUERY_HISTORY_DB_NAME);
+          db = new Database(QueryHistoryPersistor.QUERY_HISTORY_DB_NAME, QueryHistoryPersistor.QUERY_HISTORY_DB_COMMENT,
+              location, null);
+          hive.createDatabase(db, false);
+        }
+        return db;
+      } catch (HiveException e) {
+        throw new RuntimeException(e);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private String getDatabaseLocation(String databaseName) {
```

Comment:
that logic should already exists, maybe @ayushtkn could help

Reply:
refactored this area to reuse the expensive Hive object, please let me know if I can do anything else
regarding logic: did you mean creating a database + table, or specifically sys database (I was thinking about creating the sys database by HS2 in a centralized place, regardless of the query history, if it's not done yet)
cc: @ayushtkn

Reply:
I think I can remember something similar but I  just see test classes and non-test classes with incomplete usage of HIVE_METASTORE_WAREHOUSE_EXTERNAL

<img width="819" alt="Screenshot 2025-01-24 at 09 16 37" src="https://github.com/user-attachments/assets/9887dbeb-b586-4244-b765-148f9ce58e4a" />